Goenrichment (#1190)
galaxyproject/training-material: topics/transcriptomics/tutorials/goenrichment/tutorial.md

**What are GO annotations?** <br/>
Genes are associated to GO terms via GO annotations. Each gene can have multiple annotations, even of the same GO type. An important notion to take into account when using GO is that, according to the **true path rule**, a gene annotated to a term is also implicitly annotated to each ancestor of that term in the GO graph. GO annotations have evidence codes that encode the type of evidence supporting them: only a small minority of genes have experimentally verified  annotations; the large majority have annotations inferred electronically based on sequence homology or known patterns.

> ### Overview
>
> In this tutorial, we will deal with:
> 
> 1. Functional Enrichment Analysis
> 2. Methods to simplify the results
> 3. Interpretation of the results
> {:toc}
>
{: .agenda}

 
# Functional Enrichment Analysis

To perform functional enrichment analysis, we need to have:
- A set of genes of interest (e.g., differentially expressed genes): **study set**
- A set with all the genes to consider in the analysis: **population set** (which must contain the study set)
- GO annotations, associating the genes in the population set to GO terms

> - The GO ontology, with the description of GO terms and their relationships

Reviewer: I can see it in the screenshot but what about adding here, or somewhere in the text, that the ontology file needs to be in OBO or OWL format, maybe with a sentence on what the format is or a link to an explanation? e.g. http://www.geneontology.org/faq/what-obo-file-format, http://www.geneontology.org/faq/what-owl-file. And/or a link in the text to where the files come from (http://geneontology.org/page/download-ontology) could be good. Those ontology files can be used with any species can they? As maybe could be good to explicitly state that.

The appropriate statistical test is the one-tailed variant of Fisher’s exact test, also known as the hypergeometric test for over-representation. When the one-tailed version is applied, this test will compute the probability of observing at least the sample frequency, given the population frequency.  The [hypergeometric distribution](https://en.wikipedia.org/wiki/Hypergeometric_distribution) measures precisely the probability of **k** successes in **n** draws, without replacement, from a finite population of size **N** that contains exactly **K** successful objects: 
> 
![](../../images/goenrichment_formula.png)

For this first exercise we will use data from [Trapnell et al. 2014](https://www.ncbi.nlm.nih.gov/pubmed/22383036 "Trapnell et al. data"). In this work, the authors created an artificial dataset of gene expression in *Drosophila melanogaster*, where 300 random genes were set (insilico) to be differentially expressed between two conditions.

> ### {% icon hands_on %} Hands-on:
> The data for this tutorial is available at [Zenodo](https://zenodo.org/record/1255038#.Wx4qTBwh3CI) to download.
> 1. **Create a new history** 
>

> > 2. **Upload to the Galaxy** the following files:

Reviewer: Could maybe add the snippet for importing files from links?
 {% include snippets/import_via_link.md %}

Author: Thanks!

Yes, that would be good, but in this case the Zenodo link is a zip with some internal folders... when I upload this in Galaxy, it only gets the last file in the zip. I should probably make another Zenodo repository with a flat structure, and then I could import via link.

The files are relatively small in any case, and it's probably nice that the user can see them and explore them in their local computer...

>    - go.obo 
>    - drosophila_gene_association.fb 
>    - trapnellPopulation.tab 
>
>    > ### {% icon tip %} Tip: Upload data to Galaxy [<span style="color:red">[1]</span>](https://galaxyproject.github.io/training-material/topics/introduction/tutorials/galaxy-intro-peaks2genes/tutorial.html) 
>    > * **Click** on the upload button in the upper left of the interface.
>    >
>    > ![](../../images/goenrichment_galaxy_upload.png) 
>    >
>    > * Press **Choose local file** and search for your file.
>    > * Press **Start** and wait for the upload to finish.
>    {: .tip}
>
> 3. **Rename** the *go.obo* file to **GO** and *drosophila_gene_association.fb* file to **GO annotations Drosophila melanogaster**. 
>

> >    > After you upload the files, and if you press the **eye icon** of *trapnellPopulation.tab* you should look someting like this:

Reviewer: ```suggestion
>    > After you upload the files, and if you press the {% icon galaxy-eye %} (eye) icon of *trapnellPopulation.tab* you should look something like this:
```

>    > ![](../../images/goenrichment_trapnellFile.png)
> **Figure 2** Trapnell file
>
>    > As you can see, we only have the population file, containing all genes. We have to define our genes of interest.
>

> > 4. In the tool menu, navigate to `Filter and Sort -> Filter data on any column using simple expressions`.

Reviewer: I think you could omit this step (and move the comment) and just go straight to the Filter tool.

>
>    > ### {% icon comment %} Comments
>    > The study set represents the differentially expressed genes. These were chosen as having an adjusted p-value for the differential expression test (last column) smaller than a given threshold. In this case, we selected the genes with an adjusted p-value < 0,05.
>    {: .comment}
>
>    > Let's go create the study set with the help of the **Filter** tool.
> 

> > 5. **Filter** {% icon tool %}: We need to change the following settings:

Reviewer: I think the parameters are supposed to be formatted like here (I need to go change mine): https://github.com/galaxyproject/training-material/pull/1128#discussion_r232567336

```suggestion
> 4. **Filter data on any column using simple expressions** {% icon tool %} with the following parameters:
>    - {% icon param-file %} *"Filter"*: `trapnellPopulation.tab`
>    - {% icon param-text %} *"With following condition"*: `c7 < 0.05`
```

>    {: .question}
>
> 8. **Rename** files to MF Trapnell, BP Trapnell, CC Trapnell, MF graphTrapnell, BP graphTrapnell and CC graphTrapnell, respectively.
>
>
> As you can see, the output consists of a table with p-values and frequencies for each GO type plus an image with a graph view of the GO type, where you can visualize the enrichment results and highlighted enriched ontology branches. 
>
>    > ### {% icon comment %} Comments
>    > For each GO term we obtain a p-value corresponding to a single, independent test. Since we are making multiple similar tests, the probability of at least one of them being a false positive increases. Therefore we need to make a correction for multiple testing.
>    {: .comment} 
>
>    > ### {% icon question %} Questions
>    >
>    > How many significant terms do we get?
>    > 
>    > <details> 
>    > <summary>Click to view answers</summary>
>    > When we ask how many significant terms, we want to see GO terms that have a p-value < 0.05. According with the results, for Molecular Function we have 5 GO terms, Biological Process we have 43 GO terms and Component Cellular we have 10 GO terms.
>    > </details>
>    {: .question}
> 
>
>

> > If you press the eye icon of the *Molecular Function* you should see something like this:

Reviewer: ```suggestion
> If you press the {% icon galaxy-eye %} (eye) icon of the *Molecular Function* file you should see something like this:
```

> 
> ![](../../images/goenrichment_mfTrapnell.png)
>
>    > ### {% icon question %} Questions
>    >
>    > Did you expect to see significant terms?
>    > 
>    > <details> 
>    > <summary>Click to view answers</summary>

> >    > The ~300 genes should be random, wo we wouldn't expect to see any enriched term. Nonetheless we still have significant terms.

Reviewer: ```suggestion
>    > The ~300 genes should be random, so we wouldn't expect to see any enriched term. Nonetheless we still have significant terms.
```

>    > </details>
>    {: .question}
{: .hands_on}

> ### {% icon comment %} Comments
>
> Let's go back a little bit, and reopen the trapnellPopulation file. If you go through the file, you'll see genes with 'NA' as an adjusted p-value. This means that there are genes in our population for which the differential expression test was not even performed (usually genes that were not expressed in any sample). These genes are irrelevant for this functional enrichment analysis.
{: .comment} 
>
> Let's remove the irrelevant genes from trapnellPopulation.tab, to see the differences in results.
> ### {% icon hands_on %} Hands-on:
>
> 1. **Filter** {% icon tool %}: We need to change the following settings:
>    - **Filter**: `trapnellPopulation.tab`
>    - **With following condition**: `c7 != 'NA'`
> ![](../../images/goenrichment_galaxyFilterNA.png)
>

> > 2. This generate one file called *File on data 32*. **Rename** to trapnellNewPopulation.

Reviewer: It might not always be "File on data 32"?

```suggestion
> 2. This generates one file. **Rename** to `trapnellNewPopulation`.
```

  - mblue9
---

# Introduction
{:.no_toc}

When we have a large list of genes of interest (for example, a list of differentially expressed genes obtained from an RNA-Seq experiment), how do we extract biological meaning from that list?

One way is to perform functional enrichment analysis. This method consists of the application of statistical tests to verify if genes of interest are more often associated to certain biological functions than what would be expected in a random set of genes. In this tutorial you will learn about enrichment analysis and how to perform it.

**What is the Gene Ontology?** <br/>
The [Gene Ontology](http://www.geneontology.org/) (GO) is a structured, controlled vocabulary and classification of gene function at the molecular and cellular level. It is divided in three separate sub-ontologies or GO types: biological process (e.g., signal transduction), molecular function (e.g., ATPase activity) and cellular component (e.g., ribosome). These sub-ontologies are structured as directed acyclic graphs (a hierarchy with multi-parenting) of GO terms.

![](../../images/goenrichment_GOexample.png)
>
**Figure 1** QuickGO - http://www.ebi.ac.uk/QuickGO

> The GO Ontology, like other ontologies, are usually coded in the [OBO](http://www.geneontology.org/faq/what-obo-file-format) or the [OWL](http://www.geneontology.org/faq/what-owl-file) formats. It can be downloaded from the [Gene Ontology website](http://geneontology.org/page/download-ontology) or from the [OBO foundry](http://www.obofoundry.org/).

Reviewer 2: Maybe we can also mention here that there are a bunch of tools to manipulate and work with OBO files?
This one for example: https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/iuc/onto_tk_obo2owl/onto_tk_obo2owl/1.45.0
